test/evp_test.c

```diff
@@ -865,18 +865,26 @@ static int cipher_test_enc(EVP_TEST *t, int enc, size_t out_misalign,
 
     /* Test that the cipher dup functions correctly if it is supported */
     ERR_set_mark();
-    if (EVP_CIPHER_CTX_copy(ctx, ctx_base)) {
-        EVP_CIPHER_CTX_free(ctx_base);
-        ctx_base = NULL;
-    } else {
-        EVP_CIPHER_CTX_free(ctx);
-        ctx = ctx_base;
+    if (!EVP_CIPHER_CTX_copy(ctx, ctx_base)) {
+        if (fips_provider_version_ge(libctx, 3, 0, 0)) {
```

Comment:
Assuming this will be merged before 3.0.11 release - then this needs to be `fips_provider_version_ge(libctx, 3, 0, 11)` - but it is still not so easy because we have also 3.1 - i.e., this needs to be:

`(fips_provider_version_ge(libctx, 3, 0, 11) && fips_provider_version_lt(libctx, 3, 1, 0)) || fips_provider_version_ge(libctx, 3, 1, 3)`

Reply:
wouldn't it be more prudent to commit this as is, then follow up with separate PR's for each branch introducing a custom check for the right version on that branch?  That seems like it would be more understandable to me.

Reply:
No that's not sufficient.  We validate that FIPS providers are both forward and backward compatible across minor versions.

In the [provider-compatibility.yml](https://github.com/openssl/openssl/blob/master/.github/workflows/provider-compatibility.yml) unit test, we check all validated versions of the FIPS provider against a number of branches.
A (hyperthoetical) 3.0.12 FIPS provider will be tested against the latest 3.1.x branch e.g. and it will support the dupctx method & therefore must be tested.

In my experience, we should keep all of the different FIPS validated branches (3.0.x, 3.1.x and master currently) in sync with checks like this.

Reply:
nhorman FYI - the CI run that checks this is: https://github.com/openssl/openssl/actions/workflows/provider-compatibility.yml

Reply:
updated